File: partiql-plan/src/main/java/org/partiql/plan/rex/RexPivot.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.OperatorVisitor;
import org.partiql.plan.rel.Rel;
import org.partiql.types.PType;

import java.util.List;

/**
 * Logical pivot expression abstract base class.
 */
public abstract class RexPivot extends RexBase {

    /**
     * @return new RexPivot instance
     */
    @NotNull
    public static RexPivot create(@NotNull Rel input, @NotNull Rex key, @NotNull Rex value) {
        return new Impl(input, key, value);
    }

    /**
     * @return input rel (operand 0)
     */
    @NotNull
    public abstract Rel getInput();

    /**
     * @return key rex (operand 1)
     */
    @NotNull
    public abstract Rex getKey();

    /**
     * @return value rex (operand 2)
     */
    @NotNull
    public abstract Rex getValue();

    @NotNull
    @Override
    protected final RexType type() {
        return RexType.of(PType.struct());
    }

    @NotNull
    @Override
    protected final List<Operand> operands() {
        Operand c0 = Operand.single(getInput());
        return List.of(c0);
    }

    @Override
    public <R, C> R accept(OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitPivot(this, ctx);
    }

    private static class Impl extends RexPivot {

        private final Rel input;
        private final Rex key;
        private final Rex value;

        private Impl(Rel input, Rex key, Rex value) {
            this.input = input;
            this.key = key;
            this.value = value;
        }

        @NotNull
        @Override
        public Rel getInput() {
            return input;
        }

        @NotNull
        @Override
        public Rex getKey() {
            return key;
        }

        @NotNull
        @Override
        public Rex getValue() {
            return value;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/rex/RexPivot.kt
package org.partiql.plan.v1.rex

import org.partiql.plan.v1.rel.Rel
import org.partiql.types.PType

/**
 * TODO DOCUMENTATION
 */
public interface RexPivot : Rex {

    public fun getInput(): Rel

    public fun getKey(): Rex

    public fun getValue(): Rex

    public override fun getType(): PType = PType.struct()

    public override fun getOperands(): List<Rex> = listOf(getKey(), getValue())

    public override fun <R, C> accept(visitor: RexVisitor<R, C>, ctx: C): R = visitor.visitPivot(this, ctx)

    /**
     * An abstract [RexPivot] implementation intended for extension.
     */
    public abstract class Base(
        input: Rel,
        key: Rex,
        value: Rex,
    ) : RexPivot {

        // DO NOT USE FINAL
        private var _input = input
        private var _key = key
        private var _value = value

        private var operands: List<Rex>? = null
        private var type: PType? = null

        public override fun getInput(): Rel = _input

        public override fun getKey(): Rex = _key

        public override fun getValue(): Rex = _value

        public override fun getType(): PType {
            if (type == null) {
                type = PType.struct()
            }
            return type!!
        }

        public override fun getOperands(): List<Rex> {
            if (operands == null) {
                operands = listOf(getKey(), getValue())
            }
            return operands!!
        }

        public override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is RexPivot) return false

            if (_input != other.getInput()) return false
            if (_key != other.getKey()) return false
            if (_value != other.getValue()) return false

            return true
        }

        public override fun hashCode(): Int {
            var result = 1
            result = 31 * result + _input.hashCode()
            result = 31 * result + _key.hashCode()
            result = 31 * result + _value.hashCode()
            return result
        }
    }
}
